### ISSUE TITLE
Adding a task after "magento:cache:flush" is not working anymore

### ISSUE THREAD

**Issue by dverkade** [davidalger/capistrano-magento2#143]
**Description**

We have a custom task which reload PHP FPM during a deployment. This is done after the Magento cache is flushed. The task is added in the deploy.rk file with the following code: 

```
after 'magento:cache:flush', 'php:reload'
```

**Version**

Issue is on 0.9.1 and I can confirm this was working on 0.8.9 without a problem.

**Expected result**

Run the task "php:reload" after the task "magento:cache:flush is exectued"

**Actual result**

The "php:reload" task didn't run and was skipped. 

**Workaround**

When I changed the deploy.rb to the following it was working again after the mentioned task. So this is clearly something that only happens when you have a custom task after "magento:cache:flush"

```
after 'magento:maintenance:disable', 'php:reload'
```

**Comment by dverkade** [comment]
CC @bjarn

**Comment by davidalger** [comment]
Thanks for the report @dverkade. Pretty sure this is likely related to this commit ed1fa819 which was done in order to fix #138. Looks like I'm going to need to find a different solution that doesn't require calling `Rake::Task['magento:cache:flush'].reenable` which is, I'm assuming, what is breaking the ability to use `after` on this task.

**Comment by davidalger** [comment]
Alright, I made the time to test this out for myself this morning and I know what's happening here. The code I referenced is doing exactly what it's expected to do, but that also breaks your use-case of using the after hook to reload php after the deploy finishes. Right now if you look closely, you'll probably see that php:reload is being called earlier in the routine after the cache:flush which occurs immediately prior to the app:config:import:

```
03:15 magento:maintenance:check
      Checking maintenance status...
      Maintenance mode is currently inactive.

      Checking database status...
      All modules are up to date.

      Checking config status...
      Config files are up to date.

      Disabling maintenance mode management...
      There are no database updates or config changes. This is a zero-down deployment.
03:22 magento:cache:flush
      01 php -f bin/magento -- cache:flush
      01 Flushed cache types:
      01 config
      01 layout
      01 block_html
      01 collections
      01 reflection
      01 db_ddl
      01 compiled_config
      01 eav
      01 customer_notification
      01 config_integration
      01 config_integration_api
      01 full_page
      01 target_rule
      01 config_webservice
      01 translate
      01 google_product
      01 vertex
    ✔ 01 www-dev@198.101.141.127 1.565s
03:24 after:cache:flush
      after:cache:flush-executed
      01 date
      01 Fri Jun 12 08:46:06 CDT 2020
    ✔ 01 www-dev@198.101.141.127 0.321s
03:24 magento:app:config:import
      01 php -f bin/magento -- app:config:import
      01 Nothing to import.
    ✔ 01 www-dev@198.101.141.127 4.343s
03:29 magento:setup:db:schema:upgrade
      01 php -f bin/magento -- setup:db-schema:upgrade
      01 [13:46:13] Schema creation/updates:

```

It's NOT called after the second execution of magento:cache:flush however, since Rake tasks are considered "no longer needed" after being invoked and thus not executed when calling `invoke` on them a second time. The two calls to magento:cache:enable worked because of the `reenable` call I added to it making it capable of being called more than once. This reenable call doesn't however re-enable other tasks which are ascribed to it via the `after` method, resulting in the `after` method being called on the first execution of magento:cache:flush but not the second.

In order to preserve compatibility with use cases like this, I'm going to execute cache:flush directly prior to the app:config:import rather than using the invoke method, avoiding the situation but still accomplishing what was needed to resolve #138

Update coming really shortly, just have to finish testing, push, tag and release.

**Comment by davidalger** [comment]
Alright, I made the time to test this out for myself this morning and I know what's happening here. The code I referenced is doing exactly what it's expected to do, but that also breaks your use-case of using the after hook to reload php after the deploy finishes. Right now if you look closely, you'll probably see that php:reload is being called earlier in the routine after the cache:flush which occurs immediately prior to the app:config:import:

```
03:22 magento:cache:flush
      01 php -f bin/magento -- cache:flush
      01 Flushed cache types:
      01 config
      01 layout
      01 block_html
      01 collections
      01 reflection
      01 db_ddl
      01 compiled_config
      01 eav
      01 customer_notification
      01 config_integration
      01 config_integration_api
      01 full_page
      01 target_rule
      01 config_webservice
      01 translate
      01 google_product
      01 vertex
    ✔ 01 www-dev@198.101.141.127 1.565s
03:24 after:cache:flush
      after:cache:flush-executed
      01 date
      01 Fri Jun 12 08:46:06 CDT 2020
    ✔ 01 www-dev@198.101.141.127 0.321s
03:24 magento:app:config:import
      01 php -f bin/magento -- app:config:import
      01 Nothing to import.
    ✔ 01 www-dev@198.101.141.127 4.343s
03:29 magento:setup:db:schema:upgrade
      01 php -f bin/magento -- setup:db-schema:upgrade
      01 [13:46:13] Schema creation/updates:

```

It's NOT called after the second execution of magento:cache:flush however, since Rake tasks are considered "no longer needed" after being invoked and thus not executed when calling `invoke` on them a second time. The two calls to magento:cache:enable worked because of the `reenable` call I added to it making it capable of being called more than once. This reenable call doesn't however re-enable other tasks which are ascribed to it via the `after` method, resulting in the `after` method being called on the first execution of magento:cache:flush but not the second.

In order to preserve compatibility with use cases like this, I'm going to execute cache:flush directly prior to the app:config:import rather than using the invoke method, avoiding the situation but still accomplishing what was needed to resolve #138

Update coming really shortly, just have to finish testing, push, tag and release.

**Comment by dverkade** [comment]
@davidalger, thanks for debugging this one! I just checked our deployment logs and indeed it is executed, but only after the first cache flush. I didn't notice that when I was looking at this issue. See snippet of the log to see that it was executed but earlier. 

```
07:40 magento:maintenance:enable
      01 php -f bin/magento -- maintenance:enable
      01 Enabled maintenance mode
    ✔ 01 nextmp@mcenl004-fs1.nexcess.net 1.766s
      01 Enabled maintenance mode
    ✔ 01 nextmp@mcenl004-fs1.nexcess.net 2.348s
07:45 magento:cache:flush
      01 php -f bin/magento -- cache:flush
      01 Flushed cache types:
      01 config
      01 layout
      01 block_html
      01 collections
      01 reflection
      01 db_ddl
      01 compiled_config
      01 eav
      01 customer_notification
      01 config_integration
      01 config_integration_api
      01 google_product
      01 full_page
      01 config_webservice
      01 translate
      01 amasty_shopby
      01 ec_cache
      01 vertex
    ✔ 01 nextmp@mcenl004-fs1.nexcess.net 1.792s
07:47 php:reload
      01 sudo service php72u-php-fpm reload
      01 Reloading php-fpm: 
      01 [06-Jun-2020 04:48:00] NOTICE: configuration file /opt/nexcess/php72u/root/etc/php-fpm.conf test is successful
      01 
      01 [
      01   OK  ]
      01 
      01 
    ✔ 01 nextmp@mcenl004-fs1.nexcess.net 0.391s
07:47 magento:app:config:import
      01 php -f bin/magento -- app:config:import
      01 Nothing to import.
    ✔ 01 nextmp@mcenl004-fs1.nexcess.net 9.441s
```

**Comment by davidalger** [comment]
@dverkade No problem! Version 0.9.2 has been published. Let me know if it doesn't take care of the situation for you.
